test_conformance/workgroups/test_wg_scan_reduce.cpp

```diff
@@ -112,29 +163,133 @@ template <typename C> struct Reduce
+                else
+                {
+                    if (result != outptr[i + j])
+                    {
+                        log_info("%s_%s: Error at %zu\n", testName, testOpName,
+                                 i + j);
+                        return -1;
+                    }
                 }
             }
         }
         return 0;
     }
+
+    static void generate_reference_values(Type *inptr, size_t n_elems,
+                                          size_t max_wg_size, Type &max_err = 0)
+    {
+        MTdataHolder d(gRandomSeed);
+        if constexpr (std::is_floating_point_v<
+                          Type> || std::is_same_v<Type, cl_half>)
+        {
+            std::vector<Type> ref_vals(max_wg_size, 0);
+            if (std::is_same_v<Type, cl_half>)
+            {
+                // to prevent overflow limit range of randomization
+                float max_range = 99.0;
+                float min_range = -99.0;
+                // generate reference values for one work group
+                for (size_t j = 0; j < max_wg_size; j++)
+                    ref_vals[j] = cl_half_from_float(
+                        get_random_float(min_range, max_range, d),
+                        gHalfRoundingMode);
+
+                // populate reference data across all work groups
+                for (size_t i = 0; i < (size_t)n_elems; i += max_wg_size)
+                {
+                    size_t wg_size = std::min(max_wg_size, n_elems - i);
+                    memcpy(&inptr[i], ref_vals.data(), sizeof(Type) * wg_size);
+                }
+
+                if constexpr (std::is_same_v<Operation, Add<Type>>)
+                {
+                    // compute maximal summation error
+                    std::sort(ref_vals.begin(), ref_vals.end(),
+                              [](cl_half a, cl_half b) {
+                                  return std::abs(cl_half_to_float(a))
+                                      < std::abs(cl_half_to_float(b));
+                              });
+
+                    float s = 0.f;
+                    for (auto it = ref_vals.begin(); it != ref_vals.end(); ++it)
+                        s += std::abs(cl_half_to_float(*it));
+                    max_err = cl_half_from_float(
+                        fabs((max_wg_size - 1) * CL_HALF_EPSILON * s),
+                        gHalfRoundingMode);
+                }
+            }
+            else
+            {
+                double max_range = 999.0;
+                double min_range = -999.0;
+                for (size_t j = 0; j < max_wg_size; j++)
+                    ref_vals[j] = get_random_float(min_range, max_range, d);
+
+                for (size_t i = 0; i < (size_t)n_elems; i += max_wg_size)
+                {
+                    size_t work_group_size = std::min(max_wg_size, n_elems - i);
+                    memcpy(&inptr[i], ref_vals.data(),
+                           sizeof(Type) * work_group_size);
+                }
```

Comment:
I believe this means we'll be using the same input data for all work-groups.  While that's not inherently bad, it does mean we're doing redundant testing by running multiple work-groups.

Some options to consider:
1. Do nothing, these tests don't run for very long so a little redundant testing is fine.
2. Test with different random data for each work-group instead, but verify against the worst-case `max_err`.
3. Only test a single work-group.

Reply:
I'm leaning towards "do nothing", but I don't think this is consistent with the integral types.  As far as I can tell:

* This code, for the floating-point types, generates a set of random data for the "max work-group size", then stamps out the same data for each work-group, until each of the "n_elems" has a value.
* The code below, for the integral types, generates a random value for each of the "n_elems", which means that each work-group will have a different set of random data.

Is this correct?